2. Configuración del entorno

In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

3. EDA 

3.1 Carga los datos y realiza una inspección inicial (estructura, nombres de variables, tipos de datos).

In [3]:
data = pd.read_excel('C:/Users/Jagab/Desktop/Maestría_DS/Estadistica_para_DS/Entregas/Data.xlsx')
data = pd.DataFrame(data)
data.head(5)

id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB


In [5]:
data.shape

(920, 16)

In [6]:
data.columns

Index(['id', 'age', 'sex', 'dataset', 'cp', 'trestbps', 'chol', 'fbs',
       'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'],
      dtype='object')

In [7]:
data.isnull().sum()

id            0
age           0
sex           0
dataset       0
cp            0
trestbps     59
chol         30
fbs          90
restecg       2
thalch       55
exang        55
oldpeak      62
slope       309
ca          611
thal        486
num           0
dtype: int64

3.2 Identifica y clasifica las variables en categóricas (nominales, ordinales) y cuantitativas (discretas, continuas).

VARIABLES CUANTITATIVAS: Se expresan en números y permiten comparar magnitudes.

Continuas: 

    age (Edad)
    trestbps (Presión arterial en reposo)
    chol (Colesterol sérico)
    thalch (Frecuencia cardíaca máxima)
    oldpeak (Depresión del ST)


Discretas: 

    ca (Número de vasos principales)


VARIABLES CATEGORICAS: Describen cualidades o etiquetas y no se expresan numéricamente de forma natural.

Nominales:

    sex (sexo/genero)
    cp (tipo de dolor de pecho)
    fbs (nivel de azucar en sangre en ayunas)
    restecg (resultados de electrogardiogramas en reposo)
    exang (angina inducida por ejercicio)
    slope (relación entre indice de pendiente de segmento ST y capacidad del ejercicio)
    thal (talasemia)



3.3 Calcula medidas de resumen para las variables numéricas clave (e.g., age, trestbps, chol, thalach):

In [8]:
data.describe()

id         age    trestbps        chol      thalch     oldpeak  \
count  920.000000  920.000000  861.000000  890.000000  865.000000  858.000000   
mean   460.500000   53.510870  132.132404  199.130337  137.545665    0.878788   
std    265.725422    9.424685   19.066070  110.780810   25.926276    1.091226   
min      1.000000   28.000000    0.000000    0.000000   60.000000   -2.600000   
25%    230.750000   47.000000  120.000000  175.000000  120.000000    0.000000   
50%    460.500000   54.000000  130.000000  223.000000  140.000000    0.500000   
75%    690.250000   60.000000  140.000000  268.000000  157.000000    1.500000   
max    920.000000   77.000000  200.000000  603.000000  202.000000    6.200000   

               ca         num  
count  309.000000  920.000000  
mean     0.676375    0.995652  
std      0.935653    1.142693  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      0.000000    1.000000  
75%      1.000000    2.000000  
max      3.000000    4.000000

In [9]:
#Definición de función que calcula cada medida estadística.
def resumen(data, var):
    media = data[var].mean()
    mediana = data[var].median()
    std = data[var].std()
    qone = data[var].quantile(0.25)
    qtwo = data[var].quantile(0.50)
    qthree = data[var].quantile(0.75)
    qfour = data[var].quantile(1)
    iqr = qthree-qone
    coef_var = std/media
    resultado = {
        "Media" : media,
        "Mediana" : mediana,
        "Std" : std,
        "Q1" : qone,
        "Q2" : qtwo,
        "Q3" : qthree,
        "Q4" : qfour,
        "IQR" : iqr,
        "Coeficiente de Variación" : coef_var
    }
    return resultado

In [10]:
#Ciclo for para ejecutar función predefinida, por variable.
var = ["age","trestbps", "chol", "thalch", "oldpeak"]
for v in var:
    print(f"Resumen estadístico de {v}")
    print(resumen(data,v))  

Resumen estadístico de age
{'Media': np.float64(53.51086956521739), 'Mediana': np.float64(54.0), 'Std': np.float64(9.424685209576857), 'Q1': np.float64(47.0), 'Q2': np.float64(54.0), 'Q3': np.float64(60.0), 'Q4': np.float64(77.0), 'IQR': np.float64(13.0), 'Coeficiente de Variación': np.float64(0.176126556831418)}
Resumen estadístico de trestbps
{'Media': np.float64(132.13240418118468), 'Mediana': np.float64(130.0), 'Std': np.float64(19.066069518587476), 'Q1': np.float64(120.0), 'Q2': np.float64(130.0), 'Q3': np.float64(140.0), 'Q4': np.float64(200.0), 'IQR': np.float64(20.0), 'Coeficiente de Variación': np.float64(0.14429518358300208)}
Resumen estadístico de chol
{'Media': np.float64(199.13033707865168), 'Mediana': np.float64(223.0), 'Std': np.float64(110.7808103532305), 'Q1': np.float64(175.0), 'Q2': np.float64(223.0), 'Q3': np.float64(268.0), 'Q4': np.float64(603.0), 'IQR': np.float64(93.0), 'Coeficiente de Variación': np.float64(0.5563231197136715)}
Resumen estadístico de thalch
{'M

3.4 Tendencia central: media, mediana.

3.5 Dispersión: desviación estándar, rango intercuartílico (IQR), coeficiente de variación.

In [22]:
resumen_general = pd.DataFrame({
    "Media": data[var].mean(),
    "Mediana": data[var].median(),
    "Desviación Estándar": data[var].std(),
    "IQR": data[var].quantile(0.75) - data[var].quantile(0.25),
    "Coeficiente de Variación": data[var].std() / data[var].mean()
})

resumen_general

Media  Mediana  Desviación Estándar   IQR  \
age        53.510870     54.0             9.424685  13.0   
trestbps  132.132404    130.0            19.066070  20.0   
chol      199.130337    223.0           110.780810  93.0   
thalch    137.545665    140.0            25.926276  37.0   
oldpeak     0.878788      0.5             1.091226   1.5   

          Coeficiente de Variación  
age                       0.176127  
trestbps                  0.144295  
chol                      0.556323  
thalch                    0.188492  
oldpeak                   1.241740

In [21]:
resumen_grupo = data.groupby("num")[var].apply(
    lambda x: pd.DataFrame({
        "Media": x.mean(),
        "Mediana": x.median(),
        "Desviación Estándar": x.std(),
        "IQR": x.quantile(0.75) - x.quantile(0.25),
        "Coeficiente de Variación": x.std() / x.mean()
    })
)
resumen_grupo

Media  Mediana  Desviación Estándar     IQR  \
num                                                              
0   age        50.547445     51.0             9.433700   14.00   
    trestbps  129.913043    130.0            16.869867   20.00   
    chol      227.905612    228.0            75.832760   70.00   
    thalch    148.800512    151.0            23.608692   33.00   
    oldpeak     0.418205      0.0             0.715636    0.80   
1   age        53.528302     55.0             8.740371   11.00   
    trestbps  132.861111    130.0            19.848444   25.00   
    chol      195.255814    229.0           124.171544  122.00   
    thalch    131.035714    130.0            23.399584   35.00   
    oldpeak     1.001200      1.0             1.035416    1.80   
2   age        57.577982     58.0             7.786852   11.00   
    trestbps  133.613861    134.0            17.295069   22.00   
    chol      143.859813    193.0           131.600915  254.00   
    thalch    128.666667    130.0            22.811901   28.50   
    oldpeak     1.353465      1.5             1.209592    2.00   
3   age        59.214953     60.0             7.990594   10.50   
    trestbps  136.152174    135.0            24.842774   28.50   
    chol      159.716981    213.0           126.804894  267.75   
    thalch    120.500000    120.0            25.621207   35.50   
    oldpeak     1.581319      1.5             1.325889    1.60   
4   age        59.214286     59.0             8.283661    7.50   
    trestbps  138.720000    136.0            22.193317   30.00   
    chol      192.148148    225.0           116.627845  107.00   
    thalch    127.846154    125.0            23.944423   28.25   
    oldpeak     2.307692      2.5             1.304737    2.25   

              Coeficiente de Variación  
num                                     
0   age                       0.186631  
    trestbps                  0.129855  
    chol                      0.332738  
    thalch                    0.158660  
    oldpeak                   1.711209  
1   age                       0.163285  
    trestbps                  0.149392  
    chol                      0.635943  
    thalch                    0.178574  
    oldpeak                   1.034175  
2   age                       0.135240  
    trestbps                  0.129441  
    chol                      0.914786  
    thalch                    0.177295  
    oldpeak                   0.893700  
3   age                       0.134942  
    trestbps                  0.182463  
    chol                      0.793935  
    thalch                    0.212624  
    oldpeak                   0.838470  
4   age                       0.139893  
    trestbps                  0.159986  
    chol                      0.606968  
    thalch                    0.187291  
    oldpeak                   0.565386

3.6 Interpreta las diferencias de estas medidas al agrupar los datos por la variable objetivo (target, presencia o no de enfermedad).

* Edad: Muestra un incremento de la media de la edad a medida que aumenta el nivel de severidad (num/target)

Sin enfermedad= 50.55 años

Nivel 4= 59.21 años

Lo que al momento sugiere una relación positiva entre edad y severidad de enfermedad cardiaca

El coeficiente de variación se mantiene relativamente estable (~0.13–0.18).

* Presión aretrial en reposo (trestbps)

Presenta un incremento moderado en la media (129.9 en 0, 238.7 en 4)

La dispersión aumenta en niveles altos (num 3 y 4)

El coeficiente de variación se mantiene constante (0.129 0.182)

Existe una tendencia creciente en presión arterial según aumenta la severidad, aunque la variabilidad relativa es similar entre los grupos.

* Colesterol (chol)

Aqui la media no sigue un patrón claro

Se observa una desviación estandar extremadamente alta (75><126), y un IQR muy amplio (70 >< 267)

El Coeficiente de variación también es muy elevado (hasta 0.91) lo que podría significar alta dispersión y posible afectación por variables externas.

El colesterol presenta alta variabilidad y no hay una relación lineal con la severidad.

* Frecuencia Cardiaca Maxima (thalch)

Se obseva un patrón de disminución claro entre los grupos 0 y 3 (148.8 >> 120.5)

La media tiende a disminuir conforme aumenta la severidad, a excepción del cambio entre rangos 3 y 4.

El coeficiente de variación se mantiene relativamente estable (0.15 - 0.21)

Existe una relación inversa entre la frecuencia Cardiaca Máxima y la severidad de la enfermedad.

* Depresión ST Inducida (oldpeak)

Se observa un incremento progresivo de la media 0.41 >>> 2.30

La mediana también aumenta consistentemente

La desviación estandar crece, sin embargo, el coeficiente de variación disminuye en niveles altos. Esto apunta a mayor homogeneidad relativa dentro de los grupos severos.

Oldpeak muestra una asociación clara con la severidad de la enfermedad